src/man/cp.1.ronn
cp(1) -- copy files
===================

## SYNOPSIS

`cp` [-arcv] [-o <opts>] [-X <regex>] <source_file> <target_file><br>
`cp` [-arcv] [-o <opts>] [-X <regex>] <source_file> <target_dir><br>
`cp` [-arcv] [-o <opts>] [-X <regex>] <source_file>... <target_dir><br>
`cp` [-arcv] [-o <opts>] [-X <regex>] <target_dir>

## DESCRIPTION

**cp** is a data mover, used for everything from simple copies to incremental
backups. In the first form, **cp** copies the content of <source_file> to the
<target_file>, overwriting the target if it exists. In the second form,
<source_file> is copied to the <target_directory>, retaining the same name as
in the source. In the third form, multiple source files may be copied to a
target directory. In the last form, **cp** reads standard input, and File(2)
object it encounters has its contents copied to the <target_dir>. This makes it
possible to use **cp** as the last component of a file selection and filtering
pipeline.

The following options are available:

  * `-r`:
    Copy the entire subtree connected at <source_file>. In case <source_file>
    terminates with a `/`, then the contents of <source_file> will be copied to
    <target_dir>.  Otherwise, an entry for <source_file> itself will be created
    in <target_dir> and its contents recursively copied inside that directory.
    Like with the `-R` option of ls(1), files which implement readdir(2) but
    mark themselves "nodescend" will not be traversed as directories.

    read(2) is used to retrieve the contents of a file.

  * `-c`:
    Resume the copy. If the <target_file> is the same size or larger than
    <source_file>, nothing is done. If <target_file> is smaller, then **cp**
    tries to continue the read using <target_file>'s size as the starting
    offset.

    In some cases, files deliberately don't have a size exported by the
    native API (e.g. Google Docs). Such files are compared by modification
    time; if the source is older than the target, it is skipped.

  * `-v`:
    Verbose reporting. Writes a line for every file processed to standard
    error, indicating whether the file was copied or skipped, how much was
    copied and the reason for skipping. It is strongly recommended to use this
    option while testing large recursive copies.

  * `-X` <regex>:
    Regular expression of paths to exclude during a recursive copy. <regex>
    is a Javascript regular expression as defined in regexp(3).

  * `-o` <opts>:
    Option string in the options(7) format. See [OPTIONS][] for details.

  * `-a`:
    EXPERIMENTAL, do not use. "Archive" mode. Calls the bundle(2) operation on
    the file, which, if implemented, returns an object containing the file and
    all its accessories. This bundle object is then unbundled onto the target
    filesystem in the bdl(5) format, if supported.

    The purpose of archiving is to enable "rich" files to be migrated from
    their parent containers onto plain old filesystems. For example, `cp -a`
    of a URL will copy not only the HTML page itself, but also the embedded
    CSS, Javascript and images, so that a subsequent read of the bundle from
    the target filesystem will result in a fairly accurate representation of the
    original. Similarly, a Picasa photo bundle will contain metadata and
    thumbnails.

    **cp** is unaware of the details of bundle(2); that is the responsibility
    of the filesystem and mime type handler.

## OPTIONS

**cp** claims the `cp` namespace in options(7) strings passed through the `-o`
option.

  * `cp.chunksize`=<bytes>:
    **cp** tries to read(2) files in chunks of <bytes>. Chunk size defaults to
    4 MB. Some filesystems do not support chunked operation; in these cases
    the option is ignored.

  * `range.off`=<bytes>, `range.len`=<bytes>:
    **cp** passes these down to read(2) for reading a part of the file. Poor
    man's `dd`.

## DIAGNOSTICS

The **cp** command exits with `true` on success and non-`true` if an error
occurs.

The `-v` option results in one line per file processed printed to standard
error, indicating whether the file was copied or skipped, how much was
copied and the reason for skipping.

## EXAMPLES

Simple copy:

    $ cp http://pigshell.com/sample/oslogos.png /tmp
    $ cp /tmp/oslogos.png /mnt/anothername.png

Backup GDrive to /home:

    $ cp -rv -X '/Trash|/Shared With Me' /gdrive/username@gmail.com /home/
    cp -crv -X /Trash /gdrive/username@gmail.com /home/drivebackup 2>/home/drivebackup/cplog.$(date -f "YYYY-MM-DD-HHmmss")

Incremental backup of GDrive to /home:

    $ cp -rvc -X '/Trash|/Shared With Me' /gdrive/username@gmail.com /home/ 2>/home/cplog.$(date -f "YYYY-MM-DD-HHmmss")

Copy all spreadsheets from GDrive to Dropbox:

    $ ls /gdrive/username@gmail.com | grep -f mime spreadsheet | cp -v /dropbox/username@gmail.com

Extract third sector of a disk image:

    $ cp -vo range.off=1024,range.len=512 /home/somefile.raw /tmp/sector-3

## SEE ALSO

mv(1)


src/cmd/cp.js
/*
 * Copyright (C) 2012-2014 by Coriolis Technologies Pvt Ltd.
 * This program is free software - see the file COPYING for license details.
 */

function Cp(opts) {
    var self = this;

    Cp.base.call(self, opts);
}

inherit(Cp, Command);

Cp.prototype.usage = 'cp           -- copy files\n\n' +
    'Usage:\n' +
    '    cp [-arcv] [-o <opts>] [-X <regex>] <source_file> <target_file>\n' +
    '    cp [-arcv] [-o <opts>] [-X <regex>] <source_file> <target_directory>\n' +
    '    cp [-arcv] [-o <opts>] [-X <regex>] <source_file>... <target_directory>\n' +
    '    cp [-arcv] [-o <opts>] [-X <regex>] <target_directory>\n' +
    '    cp <-h | --help>\n\n' +
    'Options:\n' +
    '    -h --help    Show this message.\n' +
    '    -o <opts>    Options to pass to lower layers\n' + 
    '    -a           Archive metadata, resources, data in bundle format\n' +
    '    -r           Recursively descend source directory\n' +
    '    -v           Verbose reporting\n' +
    '    -c           Resume interrupted copy a la wget -c\n' +
    '    -X <regex>   Exclude paths matching regex\n';

Cp.prototype.internalUsage = 'cp           -- copy files\n\n' +
    'Usage:\n' +
    '    cp [-arcv] [-o <opts>] [-X <regex>] <file>...\n' +
    '    cp <-h | --help>\n\n' +
    'Options:\n' +
    '    -h --help    Show this message.\n' +
    '    -o <opts>    Options to pass to lower layers\n' +
    '    -X <regex>   Exclude paths matching regex\n';

Cp.prototype.next = check_next(do_docopt(pathargs(function() {
    var self = this,
        cpone = self.docopts['-r'] ? cprecursive : cpfile;

    if (self.inited) {
        return next();
    }

    /* First time */
    self.inited = true;
    self.retval = true;
    self.cliopts = optstr_parse(self.docopts['-o'], true);
    self.cpopts = self.cliopts.cp || {};
    delete self.cliopts['cp'];
    self.chunksize = self.cpopts.chunksize || (4 * 1024 * 1024);
    if (self.docopts['-X']) {
        try {
                self.xregex = new RegExp(self.docopts['-X']);
        } catch (e) {
            return self.exit(e.message);
        }
    }

    sys.lookup(self, self.target, self.cliopts, function(err, tfile) {
        if (err) {
            if (err.code !== 'ENOENT') {
                return self.exit(err, self.target);
            }
            if (self.mode === 'multi') {
                return self.exit(err, self.target);
            }
        } else {
            if (isrealdir(tfile)) {
                self.tdir = fstack_base(tfile);
                self.tdirpath = self.target;
                return next();
            } else if (self.mode === 'multi') {
                return self.exit(E('ENOTDIR'), self.target);
            }
        }
        /* mode = 'single', err = ENOENT */
        var comps = pathsplit(self.target),
            last = comps[1],
            parentdir = comps[0];
        sys.lookup(self, parentdir, self.cliopts, function(err, pfile) {
            if (err) {
                return self.exit(err, parentdir);
            } else {
                self.tdir = fstack_base(pfile);
                self.tdirpath = parentdir;
                self.tname = last;
                return next();
            }
        });
    });

    function next() {
        self.unext({}, cef(self, function(file) {
            if (file === null) {
                return self.exit(self.retval);
            }
            var spath = isstring(file) ? file : isstring(file._path) ? file._path : null,
                tname = self.tname || basename(spath);
            if (!spath) {
                return self.exit("Invalid file specification: " + file.toString());
            }

            cpone(fstack_top(self.tdir), self.tdirpath, tname, spath,
                function(err, res) {
                if (err) {
                    self.retval = false;
                }
                return next();
            });
        }));
    }

    /* Target directory object, target filename, src path */
    function cpfile(tdir, tdirpath, tname, srcpath, cb) {
        if (self.xregex && srcpath.match(self.xregex)) {
            cplog(srcpath, null, {status: 'skipped', more: 'excluded'});
            return cb(null, null);
        }
        fstat.call(self, srcpath, function(err, file) {
            if (err) {
                self.errmsg(err, srcpath);
                return cb(err);
            }
            return cpfile2(tdir, tdirpath, tname, file, function(err, res, stats) {
                cplog(srcpath, err, stats);
                return cb(err, res);
            });
        });
    }

    function simplecp(tdir, tdirpath, tname, sfile, cb) {
        var cliopts = self.cliopts;
        if (typeof sfile.read !== 'function') {
            return cb(E('ENOSYS'));
        }
        var fc = tdir.fs.constructor.fileclass,
            do_stream = fc && fc.prototype.hasOwnProperty('append') && (sfile.size > self.chunksize || cliopts.range);
        if (cliopts.range) {
            if (!do_stream) {
                return cb(E('EINVAL'));
            }
        }

        sys.lookup(self, pathjoin(tdirpath, tname), cliopts,
            function(err, res) {
            if (do_stream && (err || !self.docopts['-c'])) {
                /* in streaming, we use append. so create a 0 length file */
                var popts = $.extend({}, cliopts, {mime: sfile.mime});
                sys.putdir(self, tdir, tname, [''], popts, ef(cb, function() {
                    sys.lookup(self, pathjoin(tdirpath, tname), cliopts,
                        ef(cb, cp3));
                }));
            } else {
                return cp3(res);
            }
        });

        function cp3(tfile) {
            var opts = $.extend({}, self.cliopts, {chunksize: self.chunksize});
                
            if (tfile && self.docopts['-c'] && !opts.range) {
                if (sfile.size > 0) {
                    if (sfile.size === tfile.size) {
                        return cb(null, null, {status: 'skipped', more: 'size check'});
                    }
                    opts.range = {off: tfile.size, len: -1};
                } else if (sfile.mtime <= tfile.mtime) {
                    return cb(null, null, {status: 'skipped', more: 'mtime check'});
                }
            }
            if (do_stream) {
                var tfbase = fstack_base(tfile);
                fproc.call(self, sfile, opts, function(res, range, acb) {
                    sys.append(self, fstack_top(tfbase), res, self.cliopts,
                        ef(cb, acb));
                }, function(err, res, stats) {
                   fstack_invaldir(tdir); // Hack to get ls see updated size
                   return cb.apply(null, arguments);
                });
            } else {
                var popts = $.extend({}, self.cliopts, {mime: sfile.mime});
                sys.read(self, sfile, self.cliopts, ef(cb, function(data) {
                    sys.putdir(self, tdir, tname, [data], popts, cb);
                }));
            }
        }
    }

    /* Target directory object, target filename, src file object */
    function cpfile2(tdir, tdirpath, tname, src, cb) {
        if (self.docopts['-a'] && typeof tdir.unbundle !== 'function') {
                return cb(E('ENOSYS'));
        } else if (self.docopts['-a'] && typeof src.bundle === 'function') {
            sys.bundle(self, src, self.cliopts, ef(cb, function(data) {
                sys.unbundle(self, tdir, tname, data,
                    $.extend({}, self.cliopts, {'resume': self.docopts['-c']}), cb);
            }));
        } else if (isrealdir(src)) {
            sys.mkdir(self, tdir, tname, self.cliopts, function(err, dir) {
                if (!err || err.code === 'EEXIST') {
                    sys.lookup(self, pathjoin(tdirpath, tname), self.cliopts,
                        ef(cb, function(dir) {
                        return cb(null, fstack_base(dir));
                    }));
                } else {
                    return cb(err);
                }
            });
        } else {
            simplecp(tdir, tdirpath, tname, src, cb);
        }
    }

    function cprecursive(tdir, tdirpath, tname, src, cb) {
        if (self.xregex && src.match(self.xregex)) {
            cplog(src, null, {status: 'skipped', more: 'excluded'});
            return cb(null, null);
        }
        fstat.call(self, src, function(err, file) {
            if (err) {
                self.errmsg(err, src);
                return cb(err);
            }
            var srcbase = fstack_base(file);
            function cpr_descend(dir) {
                var s = fstack_top(srcbase);
                if (isrealdir(s)) {
                    sys.readdir(self, s, self.cliopts, ef(cb, function(files) {
                        var fnames = Object.keys(files);
                        async.forEachSeries(fnames, function(fname, acb) {
                            cprecursive(fstack_top(dir), pathjoin(tdirpath, tname), fname, pathjoin(src, fname), function(err, res) {
                                return soguard(self, acb.bind(null, null));
                            });
                        }, function(err) {
                            return cb(null, null);
                        });
                    }));
                } else {
                    return cb(null, null);
                }
            }
            if (tname === '') { // trailing slash
                return cpr_descend(tdir);
            }
            cpfile2(tdir, tdirpath, tname, file, function(err, dir, stats) {
                cplog(src, err, stats);
                return cpr_descend(dir);
            });
        });
    }

    function cpmany(tdir, tdirpath, sources) {
        var base = fstack_base(tdir);
        async.forEachSeries(sources, function(sourceFile, acb) {
            cpone(fstack_top(base), tdirpath, basename(sourceFile), sourceFile, function(err, res) {
                return soguard(self, acb.bind(null, null));
            });
        }, function(err) {
            return self.exit(err, null);
        });
    }

    function cplog(src, err, stats) {
        if (err) {
            self.errmsg(err, src);
            return;
        }
        if (!self.docopts['-v']) {
            return;
        }
        var msg = 'copied ' + src;
        if (stats) {
            if (stats.status) {
                msg = stats.status + ' ' + src;
            }
            if (stats.more) {
                msg += ' (' + stats.more + ')';
            }
            if (stats.range) {
                msg += sprintf(" (offset: %d length: %d)", stats.range.off,
                    stats.range.len);
            }
        } else {
            msg = 'copied ' + src;
        }
        self.errmsg(msg);
    }
})));

Command.register("cp", Cp);
